agreement_legal/models/agreement.py

```diff
@@ -0,0 +1,502 @@
+        self.ensure_one()
+        return self.copy(default=self._get_new_agreement_default_vals())
+
+    @api.model
+    def action_view_agreement(self, agreement):
+        return {
+            "res_model": "agreement",
+            "type": "ir.actions.act_window",
+            "view_mode": "form",
+            "view_type": "form",
+            "res_id": agreement.id,
+        }
+
+    def create_new_agreement(self):
+        res = self._create_new_agreement()
+        return self.action_view_agreement(res)
+
+    def _fill_create_vals(self, vals):
+        if vals.get("code", _("New")) == _("New"):
+            vals["code"] = self.env["ir.sequence"].next_by_code("agreement") or _("New")
+        if not vals.get("stage_id"):
+            vals["stage_id"] = self._get_default_stage_id()
+        return vals
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        return super().create([self._fill_create_vals(vals) for vals in vals_list])
+
+    # Increments the revision on each save action
+    def write(self, vals):
+        res = True
+        for rec in self:
+            has_revision = False
+            if "revision" not in vals:
+                vals["revision"] = rec.revision + 1
+                has_revision = True
+            res = super(Agreement, rec).write(vals)
+            if has_revision:
+                vals.pop("revision")
+        return res
+
+    def copy(self, default=None):
+        """Assign a value for code is New"""
+        default = dict(default or {})
+        if not default.get("code", False):
+            default.setdefault("code", _("New"))
+        res = super().copy(default)
+        res.sections_ids.mapped("clauses_ids").write({"agreement_id": res.id})
+        return res
+
+    def _exclude_readonly_field(self):
+        return ["stage_id"]
+
+    @api.model
+    def get_view(self, view_id=None, view_type=False, **options):
+        res = super().get_view(view_id, view_type, **options)
+        # Readonly fields
+        if view_type == "form":
+            doc = etree.XML(res["arch"])
+            for node in doc.xpath("//field"):
+                if node.attrib.get("name") in self._exclude_readonly_field():
+                    continue
+                attrs = ast.literal_eval(node.attrib.get("attrs", "{}"))
+                if attrs:
+                    if attrs.get("readonly"):
+                        attrs["readonly"] = ["|", ("readonly", "=", True)] + attrs[
+                            "readonly"
+                        ]
+                    else:
+                        attrs["readonly"] = [("readonly", "=", True)]
+                else:
+                    attrs["readonly"] = [("readonly", "=", True)]
+                node.set("attrs", simplejson.dumps(attrs))
+                modifiers = ast.literal_eval(
+                    node.attrib.get("modifiers", "{}")
+                    .replace("true", "True")
+                    .replace("false", "False")
+                )
```

Comment:
this will not work in 17.0 as the `attrs` modifiers in views are no longer working as in previous versions.